File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        best = 0
        bestBuy = -prices[0]
        lastSell = float('-inf')
        lastCoolDown = 0
        
        for p in prices:
            cd = max(bestBuy, lastSell, lastCoolDown)
            sell = bestBuy + p
            bestBuy = max(lastCoolDown - p, bestBuy)
            lastCoolDown, lastSell = cd, sell
        
        return max(lastSell, lastCoolDown)
                
            

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int bestCool = 0;
        int bestHold = -prices[0];
        int bestSell = INT_MIN;
        int oldSell = bestSell;
        
        for(auto p: prices){
            oldSell = bestSell;
            
            bestSell = max(bestSell, bestHold + p);
            
            bestHold = max(bestCool - p, bestHold);
            
            bestCool = max(bestCool, oldSell);
        }
        
        return max(bestSell, bestCool);
    }
};